### PR TITLE
search: cache results so clicking back button does not re-run search

### LINKED ISSUE #24523: Cache search results so clicking the back button after clicking on a search result does not execute the search again
Flow:

1. Execute a search and wait for results
3. Click a result.
4. Read the result, realize it's not what you're looking for. Click back button to back to results.

Current behavior:

Search runs again and you have to wait for results again, which depending on the search may take a long time (eg. unindexed searches). Because we don't have consistent sorting, results may not be in the same order as before.

Suggested behavior:

Results for the last query are cached so going back to the results page shows the same results. This should only happen if the search completed successfully.

### PR DESCRIPTION
**Pull request by limitedmage**
Fixes #24523

Uses React context to cache results based on the options passed to the search function. If the results page is loaded again with the same options, the cached results are shown instead. This significantly improves user experience when navigating results, as they can click the back button and go back to the exact same result list as before.

**Note**: scroll position is not restored when clicking back. I'm looking for a solution to this and might update this PR or send a follow-up PR to address this.

_Review tip: enable "ignore whitespace"_

### CODE REVIEW

File: client/web/src/search/results/SearchResultsCacheProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            return streamSearch(options).pipe(throttleTime(500, undefined, { leading: true, trailing: true }))
+        }, [cachedResults?.options, cachedResults?.results, options, streamSearch])
+    )
+
+    if (results?.state === 'complete') {
+        // Must be set asynchronously to avoid React errors
+        // of updating the state of a component inside another
+        // component's render function.
+        setTimeout(() => setCachedResults({ results, options }), 0)
+    }
```
Comment (tjkandala) [inline client/web/src/search/results/SearchResultsCacheProvider.tsx:51]: Would `useEffect` not work properly here?

**Review by tjkandala — APPROVED** [review]
Nice!

**Comment by 5h1rU** [comment]
Question: Are those results coming from a server request? In that case, if you replace the request logic with Apollo, you can get the cached data and (almost sure) the scroll position out of the box.

**Comment by limitedmage** [comment]
> Question: Are those results coming from a server request? In that case, if you replace the request logic with Apollo, you can get the cached data and (almost sure) the scroll position out of the box.

The request is coming from the server but not via the GraphQL API. This uses HTML5 Sever Sent Events instead so we can stream the results from the server.

**Comment by vovakulikov** [comment]
@limitedmage About scroll position restoring, I'm not sure about the current stage of that but I remember that history API has introduced something for it 
https://html.spec.whatwg.org/multipage/history.html#scroll-restoration-mode

File: client/web/src/search/results/SearchResultsCacheProvider.tsx (hunk 1 above)
Comment (limitedmage) [reply client/web/src/search/results/SearchResultsCacheProvider.tsx:51]: YES, that's much better :)

**Comment by limitedmage** [comment]
> @limitedmage About scroll position restoring, I'm not sure about the current stage of that but I remember that history API has introduced something for it
> https://html.spec.whatwg.org/multipage/history.html#scroll-restoration-mode

I don't know why it's not "just working", I assume it has something to do with either React Router or the way we have the fixed toolbar at the top with everything else scrolling with `overflow`. Anyway, I researched a bunch of solutions without success, and I think this is beyond the scope of the Search Product team, as this is an issue throughout the web application. I've filed the following bug for Frontend Platform: #24758

**Comment by limitedmage** [comment]
@sourcegraph/frontend-platform Can you chime in on the use of React context here for providing the caching functionality? The cache needs to live outside the router and I found this this be the simplest way to build it, but because of other conversations that are going on around React context, just wanted to double-check that this is not using an anti-pattern.

**Pull request merged**